aleo-programs/staking_lite/src/main.leo

```diff
@@ -0,0 +1,376 @@
+
+    // Determine if the withdrawal can fit into the current batch
+    let current_batch: u32 = current_batch_height.get_or_use(0u8, 0u32);
+    let min_claim_height: u32 = block.height + UNBONDING_PERIOD;
+    let new_batch: bool = current_batch == 0u32;
+    let unbonding_allowed: bool = new_batch || current_batch >= min_claim_height;
+    assert(unbonding_allowed);
+
+    // Assert that they have enough to withdraw
+    let delegator_balance: u64 = delegator_shares.get(owner);
+    assert(delegator_balance >= withdrawal_shares);
+
+    // Distribute shares for new commission
+    let bonded: u64 = 0u64; // credits.aleo/bonded.get(CORE_PROTOCOL);
+    let current_balance: u64 = total_balance.get(0u8);
+    let current_shares: u64 = total_shares.get(0u8);
+    let rewards: u64 = bonded - current_balance;
+    let commission_rate: u128 = commission_percent.get(0u8);
+    let new_commission: u64 = rewards > 0u64 ? get_commission(rewards as u128, commission_rate) : 0u64;
+    current_balance += rewards - new_commission;
+
+    let new_commission_shares: u64 = calculate_new_shares(current_balance as u128, new_commission as u128, current_shares as u128);
+    let current_commission: u64 = delegator_shares.get_or_use(ADMIN, 0u64);
+    delegator_shares.set(ADMIN, current_commission + new_commission_shares);
+
+    current_shares += new_commission_shares;
+    current_balance += new_commission;
+
+    // Calculate withdrawal amount
+    let withdraw_ratio: u128 = (withdrawal_shares as u128 * PRECISION_UNSIGNED) / current_shares as u128;
+    let withdraw_calculation: u128 = (current_balance as u128 * withdraw_ratio) / PRECISION_UNSIGNED;
+
+    // If the calculated withdrawal amount is greater than total_withdrawal, the excess will stay in the pool
+    assert(withdraw_calculation >= total_withdrawal as u128);
+
+    // Update withdrawals mappings
+    let batch_height: u32 = new_batch ? get_new_batch_height(block.height) : current_batch;
+    current_batch_height.set(0u8, batch_height);
+    let withdrawal: withdrawal_state = withdrawal_state {
+      microcredits: total_withdrawal,
+      claim_block: batch_height
+    };
+    withdrawals.set(owner, withdrawal);
+
+    // Update pending withdrawal
+    let currently_pending: u64 = pending_withdrawal.get(0u8);
+    pending_withdrawal.set(0u8, currently_pending + total_withdrawal);
+
+    // Update total balance
+    total_balance.set(0u8, current_balance - total_withdrawal);
+
+    // Update total shares
+    total_shares.set(0u8, current_shares - withdrawal_shares);
+
+    // Update delegator_shares mapping
+    delegator_shares.set(owner, delegator_balance - withdrawal_shares);
+  }
+
+  inline get_new_batch_height(height: u32) -> u32 {
+    let rounded_down: u32 = (height) / 1_000u32 * 1_000u32;
+    let rounded_up: u32 = rounded_down + 1_000u32;
+    return rounded_up;
+  }
+
+  transition withdraw_claim(amount: u64) {
+    credits.aleo/claim_unbond_public();
```

Comment:
This can only be called if there's an [unbond state](https://github.com/AleoHQ/snarkVM/blob/01c8b095cfa8bbeb4dedc97543966af8026bce61/synthesizer/program/src/resources/credits.aleo#L650), so there should be a separate function to claim the unbond and then to withdraw a user's claim.